File: vrep_robot_commander/src/vrep_robot_commander.cpp
#include <vrep_robot_commander/vrep_robot_commander.h>

namespace vrep{

//Constructor
RobotCommander::RobotCommander()
{
       
    //Control modes for the robot joints
    m_joint_control_modes.resize(7);

    // Global variables (modified by topic subscribers):
    m_simulationRunning=true;
    m_simulationTime=0.0f;
    m_motor_handles_available = false;

    // 1. Let's subscribe to V-REP's info stream (that stream is the only one enabled by default,
    // and the only one that can run while no simulation is running):
    m_subInfo = m_nh_.subscribe("/vrep/info",1,&RobotCommander::infoCallback,this);

    //Prepare Subcriber for motor handles:
    m_subMotorHandles = m_nh_.subscribe("/vrep_ros_communication/motor_handles",1,&RobotCommander::motorhandlesCallback,this);

    //Prepare Subcriber and Publisher
    m_lbr_motor_speedPub = m_nh_.advertise<vrep_common::JointSetStateData>("/vrep_ros_communication/lbr_joints",1);   

    // Service client to get current joint positions of manipulator from Vrep
    client_joint_config = m_nh_.serviceClient<vrep_common::simRosGetJointState>("/vrep/simRosGetJointState");


    //Start the simulation mode in VREP
    ros::ServiceClient client_startSimulation=m_nh_.serviceClient<vrep_common::simRosStartSimulation>("/vrep/simRosStartSimulation");
    vrep_common::simRosStartSimulation srv_startSimulation;
    client_startSimulation.call(srv_startSimulation);
    

    //Wait until motor handles become available (published by vrep_ros_communication node)
    while(m_motor_handles_available == false)
    {
        ROS_INFO("Waiting for motor handles callback. Make sure to activate simulation in V-REP!!!");

        // handle ROS messages:
        ros::spinOnce();

        // sleep a bit:
        usleep(5000);
    }
    ROS_INFO("Robot motor handles available!");
    
    
    //Set the joint control modes (position or velocity mode)
    setInitialControlModes();
   

}

//Destructor
RobotCommander::~RobotCommander()
{
	//Nothing to do yet
}



// Topic subscriber callbacks:
void RobotCommander::infoCallback(const vrep_common::VrepInfo::ConstPtr& info)
{
    m_simulationTime=info->simulationTime.data;
    m_simulationRunning=(info->simulatorState.data&1)!=0;
}


void RobotCommander::motorhandlesCallback(const std_msgs::Int32MultiArray::ConstPtr& handles_array)
{
    //Set motor handle vector size
    m_motor_handles.resize(handles_array->data.size());

    //Get the motor handle ID's from the message
    int i = 0;
    for(std::vector<int>::const_iterator it = handles_array->data.begin(); it != handles_array->data.end(); ++it)
        {
            m_motor_handles[i] = *it;
            i++;
        }

    //Motor Handles are now set
    m_motor_handles_available = true;

}



//Set the joint control modes
bool RobotCommander::setJointControlModes(vector<int> joint_control_modes)
{
    bool modes_set = false;

    //Publish control modes
    ros::ServiceClient client_setControlModes=m_nh_.serviceClient<vrep_common::simRosSetObjectIntParameter>("/vrep/simRosSetObjectIntParameter");
    vrep_common::simRosSetObjectIntParameter setControlModes;
    //Set streamCmd
    setControlModes.request.parameter = sim_jointintparam_ctrl_enabled;

    //Set Motor control mode for Robot Joints (last handle is for the entire robot -> used to get the current robot pose)
    for (int i = 0; i < m_motor_handles.size()-1; i++)
    {
        setControlModes.request.handle = m_motor_handles[i];
        setControlModes.request.parameterValue = joint_control_modes[i];
        //Set motor control modes for current joint
        modes_set = client_setControlModes.call(setControlModes);

        //cout<<m_motor_handles[i] <<" : "<<joint_control_modes[i];
    }

	// handle ROS messages:
    ros::spinOnce();
    
    return modes_set;
}



//Set initial Robot Pose (Start Config)
void RobotCommander::setInitialRobotPose(){

	//Service to set robot pose
    ros::ServiceClient client_initialRobotPose=m_nh_.serviceClient<vrep_common::simRosSetRobotPose>("/vrep/simRosSetInitialRobotPose");
    vrep_common::simRosSetRobotPose robot_pose;
   
   
    //Initial configuration for LBR Joints -> angle in [rad]
    robot_pose.request.position.push_back(0.0);
    robot_pose.request.position.push_back(0.0);
    robot_pose.request.position.push_back(0.0);
    robot_pose.request.position.push_back(0.0);
    robot_pose.request.position.push_back(0.0);
    robot_pose.request.position.push_back(0.0);
    robot_pose.request.position.push_back(0.0);

    //Set initial pose for robot
    client_initialRobotPose.call(robot_pose);

    cout<<"Set initial robot pose..."<<endl;

    // handle ROS messages:
    ros::spinOnce();

    // sleep a bit:
    usleep(2000);
}


//Set initial Control Modes for Joints
void RobotCommander::setInitialControlModes(){

    //Set Control Modes
    //vector<int> control_modes_values; //position control enabled -> 0 no (i.e. velocity control is used), 1 yes
    
    //LBR Joints (velocity mode = 0 or position mode = 1)
    m_joint_control_modes[0] = 0;
    m_joint_control_modes[1] = 0;
    m_joint_control_modes[2] = 0;
    m_joint_control_modes[3] = 0;
    m_joint_control_modes[4] = 0;
    m_joint_control_modes[5] = 0;
    m_joint_control_modes[6] = 0;

     
    //Set the joint control modes
    setJointControlModes(m_joint_control_modes);

    // handle ROS messages:
    ros::spinOnce();
}



//Compute and execute required joint velocities
void RobotCommander::setJointSpeedsManipulator(vector<double> joint_velocities){

    //Kuka LBR Manipulator
    float desired_LBR_joint1;
    float desired_LBR_joint2;
    float desired_LBR_joint3;
    float desired_LBR_joint4;
    float desired_LBR_joint5;
    float desired_LBR_joint6;
    float desired_LBR_joint7;

    desired_LBR_joint1 = joint_velocities[0];
    desired_LBR_joint2 = joint_velocities[1];
    desired_LBR_joint3 = joint_velocities[2];
    desired_LBR_joint4 = joint_velocities[3];
    desired_LBR_joint5 = joint_velocities[4];
    desired_LBR_joint6 = joint_velocities[5];
    desired_LBR_joint7 = joint_velocities[6];

    vrep_common::JointSetStateData lbr_motor_speeds;

    //Publish the motor speeds for the joints of the Kuka LBR:
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[0]);
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[1]);
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[2]);
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[3]);
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[4]);
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[5]);
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[6]);


    //Modes (set Joint Dynamic Properties in V-Rep accordingly)
    // 0: sets the position
    // 1: sets the target position (when joint is dynamically enabled and in position control)
    // 2: sets the target velocity (when joint is dynamically enabled without position control, or when joint is in velocity mode)
    // 3: sets the maximum force/torque that the joint can exert
    lbr_motor_speeds.setModes.data.push_back(2);
    lbr_motor_speeds.setModes.data.push_back(2);
    lbr_motor_speeds.setModes.data.push_back(2);
    lbr_motor_speeds.setModes.data.push_back(2);
    lbr_motor_speeds.setModes.data.push_back(2);
    lbr_motor_speeds.setModes.data.push_back(2);
    lbr_motor_speeds.setModes.data.push_back(2);

    lbr_motor_speeds.values.data.push_back(desired_LBR_joint1);
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint2);
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint3);
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint4);
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint5);
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint6);
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint7);

    m_lbr_motor_speedPub.publish(lbr_motor_speeds);
    
    // handle ROS messages:
    ros::spinOnce();
}



bool RobotCommander::stopManipulatorMotion(){

    vrep_common::JointSetStateData lbr_motor_speeds;

    //Kuka LBR Manipulator
    float desired_LBR_joint1;
    float desired_LBR_joint2;
    float desired_LBR_joint3;
    float desired_LBR_joint4;
    float desired_LBR_joint5;
    float desired_LBR_joint6;
    float desired_LBR_joint7;

    desired_LBR_joint1 = 0.0;
    desired_LBR_joint2 = 0.0;
    desired_LBR_joint3 = 0.0;
    desired_LBR_joint4 = 0.0;
    desired_LBR_joint5 = 0.0;
    desired_LBR_joint6 = 0.0;
    desired_LBR_joint7 = 0.0;


    //Publish the motor speeds for the joints of the Kuka LBR:
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[0]);
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[1]);
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[2]);
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[3]);
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[4]);
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[5]);
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[6]);


    //Modes (set Joint Dynamic Properties in V-Rep accordingly)
    // 0: sets the position
    // 1: sets the target position (when joint is dynamically enabled and in position control)
    // 2: sets the target velocity (when joint is dynamically enabled without position control, or when joint is in velocity mode)
    // 3: sets the maximum force/torque that the joint can exert
    lbr_motor_speeds.setModes.data.push_back(2);
    lbr_motor_speeds.setModes.data.push_back(2);
    lbr_motor_speeds.setModes.data.push_back(2);
    lbr_motor_speeds.setModes.data.push_back(2);
    lbr_motor_speeds.setModes.data.push_back(2);
    lbr_motor_speeds.setModes.data.push_back(2);
    lbr_motor_speeds.setModes.data.push_back(2);

    lbr_motor_speeds.values.data.push_back(desired_LBR_joint1);
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint2);
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint3);
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint4);
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint5);
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint6);
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint7);

    //Publish motor speeds
    m_lbr_motor_speedPub.publish(lbr_motor_speeds);


    // handle ROS messages:
    ros::spinOnce();

    // sleep a bit:
    usleep(2000); //20ms -> 50Hz loop

    cout<<"Manipulator motion stopped!"<<endl;

    return true;

}


bool RobotCommander::stopManipulatorJointMotion(int joint_num)
{
    vrep_common::JointSetStateData lbr_motor_speeds;

    //Set desired joint velocity to zero
    float desired_LBR_joint = 0.0;

    //Publish the motor speeds for the joints of the Kuka LBR:
    lbr_motor_speeds.handles.data.push_back(m_motor_handles[joint_num]);

    //Modes (set Joint Dynamic Properties in V-Rep accordingly)
    // 0: sets the position
    // 1: sets the target position (when joint is dynamically enabled and in position control)
    // 2: sets the target velocity (when joint is dynamically enabled without position control, or when joint is in velocity mode)
    // 3: sets the maximum force/torque that the joint can exert
    lbr_motor_speeds.setModes.data.push_back(2);

    //Set speed in message
    lbr_motor_speeds.values.data.push_back(desired_LBR_joint);

    //Publish motor speeds
    m_lbr_motor_speedPub.publish(lbr_motor_speeds);

    // handle ROS messages:
    ros::spinOnce();

    // sleep a bit:
    usleep(2000); //20ms -> 50Hz loop

    cout<<"Motion of joint: "<<joint_num<<" stopped!"<<endl;

    return true;
}



} //end of namespace

File: vrep_robot_commander/src/vrep_robot_commander.py
#!/usr/bin/env python


import rospy
import numpy as np
import time
from vrep_common.msg import VrepInfo, JointSetStateData
from std_msgs.msg import Int32MultiArray, UInt8MultiArray, Float32MultiArray
from vrep_common.srv import * 

class RobotCommander:
	def __init__(self, name):
		# initialize the needed variables of the class
		self.name = name
		self.num_of_joints = 7
		self.motor_handles = [None] * self.num_of_joints
		self.joint_control_modes = [None] * self.num_of_joints
		self.simulation_running = True
		self.simulation_time = 0
		self.motor_handles_available = False

		# initialize the ros node for the class
		rospy.init_node(name, anonymous=True)

		# subscribe to the "/vrep/info" topic which is available even before simulation starts.
		rospy.Subscriber("/vrep/info", VrepInfo, self.vrep_info_callback)

		# subscribe to the "/vrep_ros_communication/motor_handles" topic which is availabe when the simulation is running.
		rospy.Subscriber("/vrep_ros_communication/motor_handles", Int32MultiArray, self.motor_handles_callback)

		# initialize the publisher of the manipulator joint speeds
		# JointSetStateData.msg has 3 data values: handles, setModes, values
		self.lbr_motor_speed_pub = rospy.Publisher("/vrep_ros_communication/lbr_joints", JointSetStateData, queue_size=10)

		# service client to get current joint positions of manipulator from Vrep
		# simRosGetJointState.srv takes "handle" and returns "result" and "state"
		rospy.wait_for_service("/vrep/simRosGetJointState")
		self.joint_config_client = rospy.ServiceProxy("/vrep/simRosGetJointState", simRosGetJointState)
		

		# service client to start the simulation mode in vrep
		# simRosStartSimulation takes nothing and returns "result"
		print_data_var = False
		rospy.wait_for_service("/vrep/simRosStartSimulation")
		start_simulation_client = rospy.ServiceProxy("/vrep/simRosStartSimulation", simRosStartSimulation)
		start_simulation_result = start_simulation_client()

		if print_data_var:
			rospy.loginfo("The result of start_simulation service is %s. ", start_simulation_result)

		# wait until the motor handles are available.
		while not self.motor_handles_available:
			rospy.loginfo("Waiting for the motor handles!\n Please activate the simulation in vrep.")
			# time.sleep(5)
			rospy.sleep(5)
		
		rospy.loginfo("The motor handles are available now!")

		# set the joint control modes
		# it can be in either velocity mode (0) or position mode (1)
		self.set_initial_control_modes(0)
		
		# print the joint states:		
		print_joint_state_var = False
		if print_joint_state_var:
			for i in range(self.num_of_joints):
				joint_config_resp = joint_config_client(self.motor_handles[i])
				print(joint_config_resp)

			

	# VrepInfo.msg has 4 data values: headerInfo, simulatorState, simulationTime, timeStep
	def vrep_info_callback(self, data):
		# if you want to check the data vlues of the VrepInfo message enable the print_data_var.
		print_data_var = False
		
		# update the class variables according to the values of the topic:
		self.simulation_running = bool(data.simulatorState.data)
		self.simulation_time = data.simulationTime.data

		# print the complete data of the topic
		if print_data_var:
			rospy.loginfo('The value of headerInfo is  %s.' , data.headerInfo)
			rospy.loginfo('The value of simulatorState is  %s.' , data.simulatorState.data)
			rospy.loginfo('The value of simulationTime is  %s.' , data.simulationTime.data)

			rospy.loginfo('The value of timeStep  %s.' , data.timeStep.data)

	# Int32MultiArray has 2 data values : layout, data
	def motor_handles_callback(self, data):
		# if you want to check the data vlues of the VrepInfo message enable the print_data_var.
		print_data_var = False

		# update the class variables according to the values of the topic:
		self.motor_handles = data.data
		self.motor_handles_available = True

		# print the complete data of the topic
		if print_data_var:
			rospy.loginfo('The value of layout is  %s.' , data.layout)
			rospy.loginfo('The value of data is  %s.' , data.data)
		rospy.spin()

	# set initial control modes for the manipulator joints.
	# it can be in velocity mode (0) or position mode (1)
	def set_initial_control_modes(self, chosen_mode):
		for i in range(len(self.joint_control_modes)):
			self.joint_control_modes[i] = chosen_mode
		print("initial_control_modes")
		self.set_control_modes()

	
	# set joint control modes for the manipulator joints.
	# simRosSetObjectIntParameter.srv takes "handle", "parameter", "parameterValue and returns "reslult".
	def set_control_modes(self):
		modes_set_var = True
		sim_jointintparam_ctrl_enabled = 2001
		
		for i in range(len(self.motor_handles)):
			# rospy.wait_for_service("/vrep/simRosSetObjectIntParameter")
			set_control_modes_client = rospy.ServiceProxy("/vrep/simRosSetObjectIntParameter", 
										simRosSetObjectIntParameter)
			# set the control mode for a joint
			modes_set_var = set_control_modes_client(self.motor_handles[i], sim_jointintparam_ctrl_enabled, 												self.joint_control_modes[i])
		print("control_modes")
		return modes_set_var

	
	# set the initial robot pose
	# simRosSetRobotPose.srv takes "handle", "relativeToObjectHandle", "position" and returns "result".
	def set_initial_robot_pose(self, robot_pose = None):
		print("started")
		rospy.wait_for_service("/vrep/simRosSetInitialRobotPose")
		set_initial_robot_pose_client = rospy.ServiceProxy("/vrep/simRosSetInitialRobotPose", simRosSetRobotPose)
		if robot_pose == None:
			robot_pose = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
		set_initial_robot_pose_result = set_initial_robot_pose_client(robot_pose)
		print("initial_robot_pose_result")
		rospy.spin()

	# compute and execute joint velocities
	# we use the "self.lbr_motor_speed_pub" publisher initialized in class __init__.
	# JointSetStateData.msg has 3 data values: handles, setModes, values
	# For example handles' type is Int32MultiArray which has two attributes "data" and "layout" 
	# setModes wil set the set Joint Dynamic Properties in V-Rep and it takes either of values: 1, 2 or 3
        #       1: sets the target position (when joint is dynamically enabled and in position control)
        #	2:  sets the target velocity (when joint is dynamically enabled without position control, or when joint is in velocity mode)
        # 	3: sets the maximum force/torque that the joint can exert
	def set_manipulator_joint_speed(self, joint_velocities):
		joint_speed_msg = JointSetStateData()
		joint_speed_msg.handles.data = self.motor_handles
		joint_speed_msg.setModes.data = [2, 2, 2, 2, 2, 2, 2]
		joint_speed_msg.values.data = joint_velocities

		self.lbr_motor_speed_pub.publish(joint_speed_msg)
		print("start moving the manipulator according to the velocity!")

	# compute and execute joint angle positions
	# we use the "self.lbr_motor_speed_pub" publisher initialized in class __init__.
	# JointSetStateData.msg has 3 data values: handles, setModes, values
	# For example handles' type is Int32MultiArray which has two attributes "data" and "layout" 
	# setModes wil set the set Joint Dynamic Properties in V-Rep and it takes either of values: 1, 2 or 3
        #       1: sets the target position (when joint is dynamically enabled and in position control)
        #	2:  sets the target velocity (when joint is dynamically enabled without position control, or when joint is in velocity mode)
        # 	3: sets the maximum force/torque that the joint can exert
	def set_manipulator_joint_position(self, joint_positions):
		joint_position_msg = JointSetStateData()
		joint_position_msg.handles.data = self.motor_handles
		joint_position_msg.setModes.data = [1, 1, 1, 1, 1, 1, 1]
		joint_position_msg.values.data = joint_positions

		self.lbr_motor_speed_pub.publish(joint_position_msg)
		print("start moving the manipulator according to the position!")


	# stop the manipulator's motion
	def stop_manipulator_motion(self):
		joint_speed_msg = JointSetStateData()
		joint_speed_msg.handles.data = self.motor_handles
		joint_speed_msg.setModes.data = [2, 2, 2, 2, 2, 2, 2]
		joint_speed_msg.values.data = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]

		self.lbr_motor_speed_pub.publish(joint_speed_msg)
		print("manipulator stoped!")
		rospy.spin()


	# stop the manipulator's joint motion
	def stop_joint_motion(self, joint_num):
		joint_speed_msg = JointSetStateData()
		joint_speed_msg.handles.data = [self.motor_handles[joint_num]]
		joint_speed_msg.setModes.data = [2]
		joint_speed_msg.values.data = [0.0]

		self.lbr_motor_speed_pub.publish(joint_speed_msg)
		print("joint %d stoped!"% joint_num)
		rospy.spin()
	
	
	# print the joint states:
	def print_joint_states(self):
		print("joint information printing started!")		
		for i in range(self.num_of_joints):
			joint_config_resp = self.joint_config_client(self.motor_handles[i])
			print(joint_config_resp.state.name, ":  position is ", joint_config_resp.state.position, " velocity is ", 					joint_config_resp.state.velocity)	
		print("joint information printing ended!")

	# get the joint velocities:
	def get_joint_velocities(self):
		current_joint_velocities = np.array([0.0] * self.num_of_joints)
		for i in range(self.num_of_joints):
			joint_config_resp = self.joint_config_client(self.motor_handles[i])
			current_joint_velocities[i] = joint_config_resp.state.velocity[0]

		print "the current joint velocities are: ", current_joint_velocities
		return current_joint_velocities

	# get the joint positions:
	def get_joint_positions(self):
		current_joint_positions = np.array([0.0] * self.num_of_joints)
		for i in range(self.num_of_joints):
			joint_config_resp = self.joint_config_client(self.motor_handles[i])
			current_joint_positions[i] = joint_config_resp.state.position[0]

		# print "the current joint positions are: ", current_joint_positions
		return current_joint_positions




		


		

		
